### PR TITLE
Improvements to the scope delegation section

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 1]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
-lookup in detail. It's ok to skip this section at in the beginning.
+lookup in detail.
 To summarize what we've learned so far:
 
-- A particular scope consists of three scope axes: Subprojects, Dependency configuration, and Tasks.
-  You can think of axes like as digits in a number, or (r, g, b) components in a color.
-- There's a special value `Global` scope (written as `*` in shell) that can be applied to either of the axes.
-- There's a special value `ThisBuild` scope (written as `{.}` in shell) only for Subprojects axis.
-- build.sbt can reference a scoped key using `.in(...)` method.
+- Scopes have three axes: a subproject axis, a dependency configuration axis, and a task axis.
```
Comment (eed3si9n) [inline src/reference/00-Getting-Started/07A-Scopes.md]: You don't like my pixel analogy?

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 2]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+subproject axis, dependency configuration axis, and task axis.
 
-What is the expected `name in projB` value in the following build?
+Let's look at an example. Given the following build definition:
 
 ```scala
-scalaVersion in (ThisBuild, packageBin) := "2.12.2"
+val bippy = settingKey[String]("The biddy identifier")
+
+bippy in Global := "abc"
```
Comment (eed3si9n) [inline src/reference/00-Getting-Started/07A-Scopes.md]: It should be `bippy in (ThisBuild, packageBin) := "abc"` here.
In general build users should not use `Global` scoping, and without `packageBin` scoping we are not demonstrating the competition of two axes.

File: src/reference/00-Getting-Started/07A-Scopes.md (hunk 1 above)
Comment (dwijnand) [reply src/reference/00-Getting-Started/07A-Scopes.md]: Not entirely - I didn't feel it helped explained the concept. But I can bring it back if you feel strongly about it.

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 3]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+configuration scopes have higher precendence to settings defined in `Global` dependency configuration scope.
+
+##### Rule 2: Subproject scope axis delegation
+
+- Rule 2: For the subproject axis, sbt will delegate settings scoped in the target subproject, then to
+    `ThisBuild`, and finally fall back to `Global`.
+
+Here's an example:
+
+```scala
```
Comment (eed3si9n) [inline src/reference/00-Getting-Started/07A-Scopes.md:467]: I'm generally confused about this example. I don't think it even compiles because of the use of infix.

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 4]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+val projD2 = project dependsOn projD1
+```
+
+What is the value of `scalacOption in projD1`?
+
+* `List("-Ywarn-unused-import")`
+* `List()`
+* something else?
+
+The answer is `List("-Ywarn-unused-import")`. The behaviour is as such:
```
Comment (eed3si9n) [inline src/reference/00-Getting-Started/07A-Scopes.md]: I don't think this is true.

```
root> show projD1/scalacOptions
[info] *
```

**Review by eed3si9n — CHANGES_REQUESTED** [review]

File: src/reference/00-Getting-Started/07A-Scopes.md (hunk 1 above)
Comment (eed3si9n) [reply src/reference/00-Getting-Started/07A-Scopes.md]: Maybe I should move it up when the term axes is introduced, but I think it helps to conceptualize the idea of "scope" having multi dimensions.

File: src/reference/00-Getting-Started/07A-Scopes.md (hunk 2 above)
Comment (dwijnand) [reply src/reference/00-Getting-Started/07A-Scopes.md]: I think we should go the other way and drop the `packageBin` in the other option: we're showing how being project-scoped has higher precedence than Global-scoped. We shouldn't involve task-scoping yet, imo.

I wanted to avoid using `ThisBuild` because we've not yet introduced (at this point in the guide) how `ThisBuild` interacts in the precedence rules. So, in order to use `Global` instead, I created a brand new key, so I could use `Global` to set its default, fallback value. I'll add a comment to that effect.

File: src/reference/00-Getting-Started/07A-Scopes.md (hunk 3 above)
Comment (dwijnand) [reply src/reference/00-Getting-Started/07A-Scopes.md:467]: It was a typo instead. Thanks for catching it. (Sadly we don't have an tut or similar 😞)

**Comment by eed3si9n** [comment]
Let me hand merge some of the layout ideas presented here, and circle back to this.

File: src/reference/00-Getting-Started/07A-Scopes.md (hunk 1 above)
Comment (jvican) [reply src/reference/00-Getting-Started/07A-Scopes.md]: In my opinion, the analogy is not necessary and could even confuse beginners. It's better to leave it out and keep this section as simple as possible. The semantics are complex and analogies add a certain level of complexity for abstract terms as axis.

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 5]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+lookup in detail.
 To summarize what we've learned so far:
 
-- A particular scope consists of three scope axes: Subprojects, Dependency configuration, and Tasks.
-  You can think of axes like as digits in a number, or (r, g, b) components in a color.
-- There's a special value `Global` scope (written as `*` in shell) that can be applied to either of the axes.
-- There's a special value `ThisBuild` scope (written as `{.}` in shell) only for Subprojects axis.
-- build.sbt can reference a scoped key using `.in(...)` method.
+- Scopes have three axes: a subproject axis, a dependency configuration axis, and a task axis.
+- There's a special value `Global` (written as `*` in shell) for any of the scope axes.
```
Comment (jvican) [inline src/reference/00-Getting-Started/07A-Scopes.md]: Is it necessary to say how it's written in the shell here? I feel it belongs to another section. Also, I've also felt that these symbols are very hard to parse and internalize. I think that it would help to replace `{.}` by `ThisBuild` and`*` by `Global`.

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 6]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+    bippy := "xyz",
+    name := "foo-" + (bippy in packageBin).value
   )
 ```
 
-The answer is `foo2.11.11`. Because of Rule 1A `scalaVersion` scoped to `projB` has
-higher precendence over `(ThisBuild, packageBin)`.
+What is the value of `name in projB`?
+
+1. `foo-abc`?
```
Comment (jvican) [inline src/reference/00-Getting-Started/07A-Scopes.md]: 👍 Really good way to let people take a guess.

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 7]
```diff
@@ -379,104 +379,153 @@ lazy val util = (project in file("util"))
...
+
+```scala
+lazy val root = (project in file("."))
+  .settings(
+    inThisBuild(List(
+      // Same as:
+      // scalacOptions in ThisBuild += "-Ywarn-unused-import"
+      scalacOptions += "-Ywarn-unused-import"
+    ))
+  ) aggregate (projD1, projD2) dependsOn (projD1, projD2)
```
Comment (jvican) [inline src/reference/00-Getting-Started/07A-Scopes.md:475]: What would happen if `aggregate` and `dependsOn` are defined before `settings`?

File: src/reference/00-Getting-Started/07A-Scopes.md [hunk 8]
```diff
@@ -509,25 +558,27 @@ Hello> inspect projC/compile:console::scalacOptions
...
+Scope delegation feels similar to class inheritance in an object-oriented language,
 but there's a difference. In an OO language like Scala if there's a method named
 `drawShape` on a trait `Shape`, its subclasses can override that even when `drawShape` is used
 within the `Shape` trait, which is called dynamic dispatch.
 
-In `build.sbt`, however, if you define a setting in terms of another setting at the build level,
-a key scoped to project level might delegate to the build level, but it will *not* come back.
-Here is an example.
+In sbt, however, scope delegation can delegate to a build level setting,
+but that build level setting can't then refer to project level settings.
```
Comment (jvican) [inline src/reference/00-Getting-Started/07A-Scopes.md]: This is the rewrite i was referring to in the original PR. Excellent!

**Comment by jvican** [comment]
I really like these improvements, as well as the original text. Good work :smile:.

**Comment by eed3si9n** [comment]
Closing this but I've got most of the ideas merged by hand.
